Add custom visitor for procarg0 that supports both single argument and mlhs.

iliabylich opened the pull request:
Closes https://github.com/whitequark/parser/issues/527

@whitequark Do you see any better way to detect single argument / mlhs?

whitequark commented:
Hm, why do we emit `s(:procarg0, :a)` in the first place? Shouldn't it be `s(:procarg0, s(:arg, :a))`?

iliabylich commented:
Because we do `arg.updated(:procarg0)`, and sometimes it's a `s(:arg)`, sometimes it's a `(:mlhs, s(:arg, :a), s(:arg, :b))`

Should we fix the format of AST?

whitequark commented:
Yeah, that sounds all around better. But if it breaks user code it might need an opt-in...

iliabylich commented:
Are you talking about adding one more flag like `Builder.emit_mlhs_inside_procarg0`?

whitequark commented:
Yeah. Though we should probably not emit mlhs but only change `:x` to `(arg :x)`.

iliabylich commented:
> Though we should probably not emit mlhs but only change :x to (arg :x). 

Right. to sum up, if the flag is enabled:
+ `a` becomes `s(:procarg0, s(:arg, :a))` (that's the change)
+ `(a, b)` is still `s(:procarg0, s(:arg, :a), s(:arg, :b))`

if it's disabled we use current behavior:
+ `a` is `s(:procarg0, :a)`
+ `(a, b`) is `s(:procarg0, s(:arg, :a), s(:arg, :b))`

And then what should be the name of the flag? `Builder.emit_arg_inside_procarg0`?

And `Builder.modernize` should enable it, right?

Also it's a bit unclear what is the source map of this node. Right now it's either `Map::Collection` (if it was generated from `mlhs_node.updated(:procarg0)`) or `Map::Variable` (for `arg_node.updated(:procarg0)`). I guess it should be `Map::Collection` because it wraps 1+ arguments, but could you confirm it please?

whitequark commented:
> Right. to sum up

Correct.

> And then what should be the name of the flag? `Builder.emit_arg_inside_procarg0`?

That's fine.

> And `Builder.modernize` should enable it, right?

Correct. `modernize` enables all flags.

> I guess it should be `Map::Collection` because it wraps 1+ arguments, but could you confirm it please?

Again sounds fine.

iliabylich commented:
Done. @whitequark PTAL when you have time.

I wasn't able to find any references to `Builder` in `AST_FORMAT`, so I assume it documents a "modern" version of AST. Or it a problem with documentation? 😄

whitequark commented:
> so I assume it documents a "modern" version of AST

It does.

whitequark commented on lib/parser/ast/processor.rb:
```diff
@@ -124,7 +124,16 @@ def process_argument_node(node)
+
+      def on_procarg0(node)
+        if node.children[0].is_a?(Symbol)
+          # s(:procarg0, :a)
```
I think we don't need this branch anymore.

iliabylich replied:
But what if the builder is not "modernized"?

whitequark replied:
Then it works just like before.

But I think it's fine to leave this in. Just needs a comment to indicate why this is using an AST format that's not what our docs describe.

iliabylich commented:
@whitequark Fixed. Anything else to change/fix before merging?
